file_catalog/mongo.py

```diff
@@ -93,29 +115,34 @@ def _get_projection(
         return projection
 
     @staticmethod
-    def _limit_result_list(
-        result: List[Dict[str, Any]], limit: Optional[int] = None, start: int = 0,
+    async def _limit_result_list(
+        cursor: MotorCursor, limit: Optional[int] = None, start: int = 0,
     ) -> List[Dict[str, Any]]:
-        """Get sublist of `results` using `limit` and `start`.
+        """Get sublist of results from `cursor` using `limit` and `start`.
 
          `limit` and `skip` are ignored by __getitem__:
          http://api.mongodb.com/python/current/api/pymongo/cursor.html#pymongo.cursor.Cursor.__getitem__
 
         Therefore, implement it manually.
         """
-        ret = []
+        results = []
         end = None
 
         if limit is not None:
             end = start + limit
 
-        for row in result[start:end]:
-            ret.append(row)
+        i = 0
+        async for row in cursor:
```

Comment:
It might be better to use ` results = cursor.skip(start).limit(limit).to_list(None)`.  This lets it use server-side skipping.

Reply:
It's also short enough it might remove the need for this function :smile:

Reply:
Sure, that'd be much better. What's up with `to_list(None)`? Why `None`?

Reply:
`to_list` takes a limit parameter, for how many items it should consume before stopping.  `None` means infinite.

I suppose for a production instance you'd probably impose some upper bound for performance reasons, and if you hit that you'd want to return an error.  But since I think we're keeping FC only for in-house use, it's fine to use infinite.